src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java

```diff
@@ -27,9 +27,9 @@
 final class BoringSSLCertificateVerifyCallback {
 
     private final QuicheQuicSslEngineMap engineMap;
-    private final X509ExtendedTrustManager manager;
+    private final X509TrustManager manager;
 
-    BoringSSLCertificateVerifyCallback(QuicheQuicSslEngineMap engineMap, X509ExtendedTrustManager manager) {
+    BoringSSLCertificateVerifyCallback(QuicheQuicSslEngineMap engineMap, X509TrustManager manager) {
```

Comment:
Why this change ? This doesn't look correct as we want to pass the engine down to the manager

Reply:
The minSdkVersion of quiche is Android API 21. `X509ExtendedTrustManager` is only available on API 24+, as you can see [in the javadocs in the upper right corner](https://developer.android.com/reference/javax/net/ssl/X509ExtendedTrustManager). Increasing the minSdkVersion from 21 to 24 drops support for 20% of the devices, so I would really like to keep it at 21, especially because quiche also keeps the api at 21, I would try to avoid dropping support for 20% of the devices just for a single class.

I found [this PR](https://github.com/netty/netty/pull/5912/files) and saw that the methods we call in `BoringSSLCertificateVerifyCallback`, namely `manager.checkServerTrusted(peerCerts, authAlgorithm, engine);` and `manager.checkClientTrusted(peerCerts, authAlgorithm, engine);` are delegated to the X509TrustManager, so i just replaced the `X509ExtendedTrustManager` with `X509TrustManager` and though it was okay. But maybe I'm wrong, as you said.
Would it help to wrap the trustmanager, when possible? like seen on these lines: https://github.com/netty/netty/blob/master/handler/src/main/java/io/netty/handler/ssl/util/SimpleTrustManagerFactory.java#L139-L141
As the declarations are not public and we just call methods of the X509TrustManager, that class could be used and just wrapped on platforms where it is available, or is that a bad idea?

Reply:
These are commits from other projects with the same problem:

https://github.com/eclipse/californium/pull/1593

https://github.com/google/conscrypt/pull/645

https://github.com/BiglySoftware/BiglyBT/commit/2e0a2f9d7e20223f3ded89b5c161883abd3f978c

Reply:
https://github.com/netty/netty-incubator-codec-quic/pull/249 can you check this one...

Reply:
I will test it tomorrow evening on an old device

Reply:
Thanks a lot!

Reply:
It works! But I tried it with the change I requested on the PR (`Class.forName(X509ExtendedTrustManager.class.getName());`).

Just wanted to note that on API 21-23 I had to enable [Java 8+ API desugaring support](https://developer.android.com/studio/write/java8-support#library-desugaring), otherwise the app will crash, it won't find a lambda class. On Android API Level >=24, Java 8 features used in netty-incubator-codec-quic (like lambdas) are supported and it's working without enabling desugaring.